smartsim/_core/_install/buildenv.py

```diff
@@ -0,0 +1,355 @@
+    REDIS_URL = get_env("SMARTSIM_REDIS_URL",
+                        "https://github.com/redis/redis.git/")
+
+    # RedisAI
+    REDISAI = RedisAIVersion(get_env("SMARTSIM_REDISAI", "1.2.3"))
+    REDISAI_URL = get_env("SMARTSIM_REDISAI_URL",
+                          "https://github.com/RedisAI/RedisAI.git/")
+
+    # ML/DL (based on RedisAI version defaults)
+    TORCH = Version_(get_env("SMARTSIM_TORCH", REDISAI.torch))
+    TORCHVISION = Version_(get_env("SMARTSIM_TORCHVIS", REDISAI.torchvision))
+    TENSORFLOW = Version_(REDISAI.tensorflow)
+    ONNX = Version_(REDISAI.onnx)
+
+    def as_dict(self):
+        packages = ["SMARTSIM", "SMARTREDIS", "REDIS",
+                    "REDISAI", "TORCH", "TENSORFLOW", "ONNX"]
+        versions = [self.SMARTSIM, self.SMARTREDIS, self.REDIS,
+                    self.REDISAI, self.TORCH, self.TENSORFLOW, self.ONNX]
+        vers = {"Packages": packages, "Versions": versions}
+        return vers
+
+    def ml_extras_required(self):
+        """Optional ML/DL dependencies we suggest for the user.
+
+        The defaults are based on the RedisAI version
+        """
+        ml_extras = []
+        ml_defaults = self.REDISAI.get_defaults()
+
+        # remove torch and torch vision as they will be installed
+        # by the cli process for use in the RAI build. We don't install
+        # them here as the user needs to decide between GPU/CPU. All other
+        # libraries work on both devices
+        del ml_defaults["torch"]
+        del ml_defaults["torchvision"]
+
+        for lib, vers in ml_defaults.items():
+            ml_extras.append(f"{lib}=={vers}")
+        return ml_extras
+
+    def get_sha(self, setup_py_dir) -> str:
+        """Get the git sha of the current branch"""
+        try:
+            sha = (
+                subprocess.check_output(["git", "rev-parse", "HEAD"], cwd=setup_py_dir)
+                .decode("ascii")
+                .strip()
+            )
+            return sha[:7]
+        except Exception:
+            return "Unknown"
+
+    def write_version(self, setup_py_dir):
+        """
+        Write version info to version.py
+
+        Use git_sha in the case where smartsim suffix is set in the environment
+        """
+        version = self.SMARTSIM
+        if self.SMARTSIM_SUFFIX:
+            git_sha = self.get_sha(setup_py_dir)
+            version = f"{version}-{self.SMARTSIM_SUFFIX}-{git_sha}"
+
+        version_file = setup_py_dir / "smartsim" / "version.py"
+        with open(version_file, "w") as f:
+            f.write("# This file is automatically generated by setup.py\n")
+            f.write("# do not edit this file manually.\n\n")
+
+            f.write(f"__version__ = '{version}'\n")
+        return version
+
+
+class BuildEnv:
+    """Environment for building third-party dependencies"""
+
+    # Compiler overrides
+    CC = os.environ.get("CC", "gcc")
+    CXX = os.environ.get("CXX", "g++")
```

Comment:
how should we do that? List all available compilers? I do realize this could be a bit of a footgun, but compiler wrappers, new compilers, special platform compilers (``nvcc``), and others would be tough to exhust and maintain a list of.

I do think we should check the versions of prereqs though (i.e. ``make`` and ``cmake`` and what not).

Reply:
That's a good point. If we make a list of all available compilers then this would run the risk of accidentally missing some. 
Checking versions of prerequisites should be done though.